src/Transform/BundleMemoryPools.cpp

```diff
@@ -473,6 +473,28 @@ class KrnlBundleDynamicMemoryPools : public OpRewritePattern<AllocOp> {
   }
 };
 
+/*
+ * Move all constants to the top of their respective block to avoid
+ * unwanted merges.
+ */
+class KrnlMoveConstantsUp : public OpRewritePattern<ConstantOp> {
+public:
+  using OpRewritePattern<ConstantOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(
+      ConstantOp constOp, PatternRewriter &rewriter) const override {
+    // Get parent block.
+    auto parentBlock = constOp.getOperation()->getBlock();
+
+    // Ensure it's the top block.
+    if (!llvm::dyn_cast_or_null<FuncOp>(parentBlock->getParentOp()))
```

Comment:
Will this condition hold for nested Loop or If?

Reply:
At the moment this is only enabled for the main block since I've only observed the error there.